src/algebra/module/dedekind_domain.lean

```diff
@@ -55,15 +55,25 @@ begin
 end
 
 /--A finitely generated torsion module over a Dedekind domain is an internal direct sum of its
-`p i ^ e i`-torsion submodules for some prime ideals `p i` and numbers `e i`.-/
+`p i ^ e i`-torsion submodules where `p i` are factors of `(⊤ : submodule R M).annihilator` and
+`e i` are their multiplicities. -/
 theorem is_internal_prime_power_torsion [module.finite R M] (hM : module.is_torsion R M) :
-  ∃ (P : finset $ ideal R) [decidable_eq P] [∀ p ∈ P, prime p] (e : P → ℕ),
-  by exactI direct_sum.is_internal (λ p : P, torsion_by_set R M (p ^ e p : ideal R)) :=
+  direct_sum.is_internal (λ p : (factors (⊤ : submodule R M).annihilator).to_finset,
+    torsion_by_set R M (p ^ (factors (⊤ : submodule R M).annihilator).count p : ideal R)) :=
```

Comment:
Do you think it makes sense to turn this construction into a `def`? It looks justified to me because `p` appears twice in `torsion_by_set R M (p ^ (factors (⊤ : submodule R M).annihilator).count p`.

Reply:
Which construction are you referring to? `factors (⊤ : submodule R M).annihilator`?

Reply:
I was thinking of `torsion_by_set R M (p ^ (factors (⊤ : submodule R M).annihilator).count p : ideal R)` itself.

Reply:
I'm not sure that's all that useful. Note that `factors (⊤ : submodule R M).annihilator` appears twice in the statement too.

Probably there is a more suitable way to express the indexing than the current way (perhaps by indexing by all possible ideals, and letting the count be zero on the ones that aren't factors.

Either way, I'm not all that familiar with the math, so would prefer to make no changes and let someone else come along and think "hey, this indexing is weird". At least now it's part of the definition so it's visibly weird.